lib/Analysis/LazyRelinAnalysis/LazyRelinAnalysis.h

```diff
@@ -0,0 +1,36 @@
+#ifndef LIB_ANALYSIS_LAZYRELINANALYSIS_H
+#define LIB_ANALYSIS_LAZYRELINANALYSIS_H
+
+#include "llvm/include/llvm/ADT/DenseMap.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/Operation.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/Value.h"      // from @llvm-project
+
+namespace mlir {
+namespace heir {
+class LazyRelinAnalysis {
+ public:
+  LazyRelinAnalysis(Operation *op) : opToRunOn(op){};
+  ~LazyRelinAnalysis() = default;
+
+  LogicalResult solve();
+
+  // Return true if a relin op should be inserted after the given
+  // operation, according to the solution to the optimization problem.
+  bool shouldInsertRelin(Operation *op) const { return solution.lookup(op); }
+
+  // Return the key basis degree at the given SSA value, as determined
+  // by the solution to the optimization problem.
+  // Don't need an "after relin" version because after relin is always 1.
```

Comment:
another place that might benefit from making "key basis degree" meaning more explicit.

```suggestion
  // Don't need an "after relin" version because after relin the keybasis is always (1, s),
  // so the key basis degree is always 1
```